### DISCUSSION TITLE
Strategy for Inserting 100 Million Documents into Vector Database

### DISCUSSION THREAD

**Discussion by MartinMashalov** [milvus-io/milvus#32754] [Q&A and General discussion]
Hi there!

I am trying to insert 100 million documents into the Mulvis vector database. I am using openai embeddings for the documents. Since I am new to Mulvis, I am wondering what the best strategy would be to insert all of these documents efficiently and also as quickly as possible. Would batch indexing be a good idea? Has anyone else encountered a similar problem and figured out how to overcome it? I would appreciate any guidance on this issue. 

Thank you.

**Comment by yhmo** [comment]
We recommend batch insert for Milvus.
The RPC transfer size limit is 64MB for each insert call. So, it is better to insert data batch by batch with each batch size between 20~40MB.
Each dimension is a float32 value. So, for 1536-dim embeddings, you can insert 3000 rows ~ 7000 rows for each batch. If there are some other metadata along with the embeddings, especially long-length strings, you might reduce the row count to 1000 ~ 2000 rows for each batch.

**Comment by MartinMashalov** [comment]
Thank you for your response! How can I do this batch insert (from JSON file or through some Python integration)? Any reference to documentation would be super helpful. 

Thank you so much! 

Martin 

>

**Reply by yhmo** [reply]
The User Guide of online doc is a good start: 
Create collection: https://milvus.io/docs/manage-collections.md
Insert/delete: https://milvus.io/docs/insert-update-delete.md
Index: https://milvus.io/docs/index-vector-fields.md

After read the doc, try some examples:
https://github.com/milvus-io/pymilvus/blob/master/examples/hello_milvus.py
https://github.com/milvus-io/pymilvus/blob/master/examples/example.py


Milvus insert() interface supports both row-based and column-based insertion:
Assume we have a collection with a primary key field and a vector field with name "id" and "embedding".
Row-based insert is like this:
`collection.insert(data = {"id": 100, "embedding": [1.2, 1.3, 1.4, 1.5, 1.6.....]})`

Column-based insert is like this:
```
collection.insert(data = [
   [1, 2, 3, 4, .... 100], # primary keys
   [[1.1, 1.2, ...], [2.1, 2.2, ...], [3.1, 3.2, ...], .... [100.1, 100.2, ....]],  # embedding field
]
)

```

**Comment by Rishi20-03-syfer** [comment]
[@MartinMashalov how do you get he 100m document. Because I also need to feed vector data into my database.I appreciate if you can help me with this.

**Reply by MartinMashalov** [reply]
Yep. The bottleneck is the speed at which you embed your documents not Milvus as a database. If you're embedding locally, then consider 2x T4 GPUs. Embedding through an api (like openai) is more efficient and manages to perform embedding for 100m documents in about 1-2 days.
